File: sandbox-vercel/src/handlers.ts
import type { VercelClient } from './client.js'
import type { Config } from './config.js'
import { imageAllowed } from './config.js'
import { CAPABILITIES, S, SandboxError } from './sandbox.js'

export interface HandlerCtx {
  config: Config
  client: VercelClient
  inFlight: { value: number }
}

function requireStr(input: Record<string, unknown>, key: string): string {
  const v = input[key]
  if (typeof v !== 'string' || v.length === 0) {
    throw new SandboxError(S.PROVIDER_UNAVAILABLE, `missing string field "${key}"`)
  }
  return v
}

export async function doCreate(ctx: HandlerCtx, input: Record<string, unknown>) {
  const image = requireStr(input, 'image')
  if (!imageAllowed(ctx.config, image)) {
    throw new SandboxError(S.IMAGE_NOT_ALLOWED, `image not in allowlist: ${image}`)
  }
  const idle =
    typeof input.idle_timeout_secs === 'number' ? input.idle_timeout_secs : ctx.config.default_idle_timeout_secs
  const runtime = typeof input.runtime === 'string' ? input.runtime : ctx.config.default_runtime

  if (ctx.inFlight.value >= ctx.config.max_concurrent_sandboxes) {
    throw new SandboxError(S.CONCURRENCY_CAP, `concurrency cap reached (${ctx.inFlight.value} active)`)
  }
  ctx.inFlight.value += 1
  try {
    const sourceUrl = typeof input.source_url === 'string' ? input.source_url : undefined
    const sourceRev = typeof input.source_revision === 'string' ? input.source_revision : undefined
    const created = await ctx.client.create({
      image,
      idle_timeout_secs: idle,
      runtime,
      source: sourceUrl ? { url: sourceUrl, revision: sourceRev } : undefined,
    })
    return {
      sandbox_id: created.sandbox_id,
      image: created.image,
      started_at: created.started_at,
      capabilities: CAPABILITIES,
    }
  } catch (e) {
    ctx.inFlight.value -= 1
    throw e
  }
}

export async function doExec(ctx: HandlerCtx, input: Record<string, unknown>) {
  const sandbox_id = requireStr(input, 'sandbox_id')
  const cmd = requireStr(input, 'cmd')
  const args = Array.isArray(input.args) ? (input.args as string[]) : []
  const timeout_ms = typeof input.timeout_ms === 'number' ? input.timeout_ms : undefined
  return ctx.client.exec(sandbox_id, cmd, args, timeout_ms)
}

export async function doStop(ctx: HandlerCtx, input: Record<string, unknown>) {
  const sandbox_id = requireStr(input, 'sandbox_id')
  await ctx.client.stop(sandbox_id)
  if (ctx.inFlight.value > 0) ctx.inFlight.value -= 1
  return {}
}

export async function doList(ctx: HandlerCtx, _input: Record<string, unknown>) {
  // Best-effort upstream fetch. Local capacity envelope is always returned
  // so callers can still trust in_flight/cap/remaining when the provider
  // is down.
  let sandboxes: Awaited<ReturnType<VercelClient['list']>> = []
  try {
    sandboxes = await ctx.client.list()
  } catch {
    sandboxes = []
  }
  const cap = ctx.config.max_concurrent_sandboxes
  return {
    sandboxes,
    in_flight: ctx.inFlight.value,
    cap,
    remaining: Math.max(cap - ctx.inFlight.value, 0),
  }
}

export async function doSnapshot(ctx: HandlerCtx, input: Record<string, unknown>) {
  const sandbox_id = requireStr(input, 'sandbox_id')
  const snapshot_id = await ctx.client.snapshot(sandbox_id)
  return { snapshot_id }
}

export async function doExposePort(ctx: HandlerCtx, input: Record<string, unknown>) {
  const sandbox_id = requireStr(input, 'sandbox_id')
  const port = input.port
  if (typeof port !== 'number' || port < 1 || port > 65535) {
    throw new SandboxError(S.PROVIDER_UNAVAILABLE, 'invalid port')
  }
  const url = await ctx.client.exposePort(sandbox_id, port)
  return { url }
}

export async function doFsRead(ctx: HandlerCtx, input: Record<string, unknown>) {
  const sandbox_id = requireStr(input, 'sandbox_id')
  const path = requireStr(input, 'path')
  const bytes = await ctx.client.fsRead(sandbox_id, path)
  return { bytes_base64: Buffer.from(bytes).toString('base64') }
}

export async function doFsWrite(ctx: HandlerCtx, input: Record<string, unknown>) {
  const sandbox_id = requireStr(input, 'sandbox_id')
  const path = requireStr(input, 'path')
  const bytes_b64 = requireStr(input, 'bytes_base64')
  const bytes = Buffer.from(bytes_b64, 'base64')
  const mode = typeof input.mode === 'number' ? input.mode : undefined
  await ctx.client.fsWrite(sandbox_id, path, bytes, mode)
  return {}
}

File: sandbox-modal/src/handlers.py
"""Handler coroutines for every `sandbox::modal::*` function.

Each coroutine parses untyped JSON, does config-side validation
(allowlist, concurrency cap), then delegates to `ModalClient`. Failures
surface as `SandboxError` whose stringification carries the leading
`[Sxxx]` so callers can pattern-match on the S-code.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any

from .config import Config
from .modal_client import ModalClient
from .sandbox import (
    CAPABILITIES,
    S_CONCURRENCY_CAP,
    S_IMAGE_NOT_ALLOWED,
    S_PROVIDER_UNAVAILABLE,
    SandboxError,
)


@dataclass
class HandlerCtx:
    config: Config
    client: ModalClient
    in_flight: int = 0
    lock: asyncio.Lock | None = None

    def __post_init__(self) -> None:
        if self.lock is None:
            self.lock = asyncio.Lock()


def _require_str(payload: dict[str, Any], key: str) -> str:
    val = payload.get(key)
    if not isinstance(val, str) or not val:
        raise SandboxError(S_PROVIDER_UNAVAILABLE, f'missing string field "{key}"')
    return val


async def do_create(ctx: HandlerCtx, payload: dict[str, Any]) -> dict[str, Any]:
    image = _require_str(payload, "image")
    if not ctx.config.image_allowed(image):
        raise SandboxError(S_IMAGE_NOT_ALLOWED, f"image not in allowlist: {image}")

    cpus = int(payload.get("cpus") or ctx.config.default_cpus)
    memory_mb = int(payload.get("memory_mb") or ctx.config.default_memory_mb)
    idle = int(payload.get("idle_timeout_secs") or ctx.config.default_idle_timeout_secs)

    assert ctx.lock is not None
    async with ctx.lock:
        if ctx.in_flight >= ctx.config.max_concurrent_sandboxes:
            raise SandboxError(S_CONCURRENCY_CAP, f"concurrency cap reached ({ctx.in_flight} active)")
        ctx.in_flight += 1

    try:
        created = await ctx.client.create(image, cpus, memory_mb, idle)
        return {
            "sandbox_id": created.sandbox_id,
            "image": created.image,
            "started_at": created.started_at,
            "capabilities": list(CAPABILITIES),
        }
    except BaseException:
        async with ctx.lock:
            ctx.in_flight = max(0, ctx.in_flight - 1)
        raise


async def do_exec(ctx: HandlerCtx, payload: dict[str, Any]) -> dict[str, Any]:
    sandbox_id = _require_str(payload, "sandbox_id")
    cmd = _require_str(payload, "cmd")
    args = list(payload.get("args") or [])
    timeout_ms = payload.get("timeout_ms")
    result = await ctx.client.exec(sandbox_id, cmd, [str(a) for a in args], timeout_ms)
    return {
        "stdout": result.stdout,
        "stderr": result.stderr,
        "exit_code": result.exit_code,
        "timed_out": result.timed_out,
    }


async def do_stop(ctx: HandlerCtx, payload: dict[str, Any]) -> dict[str, Any]:
    sandbox_id = _require_str(payload, "sandbox_id")
    await ctx.client.stop(sandbox_id)
    assert ctx.lock is not None
    async with ctx.lock:
        ctx.in_flight = max(0, ctx.in_flight - 1)
    return {}


async def do_list(ctx: HandlerCtx, _payload: dict[str, Any]) -> dict[str, Any]:
    # Best-effort upstream fetch. Local capacity envelope always returned.
    sandboxes: list[dict[str, object]] = []
    try:
        sandboxes = await ctx.client.list()
    except SandboxError:
        sandboxes = []
    cap = ctx.config.max_concurrent_sandboxes
    return {
        "sandboxes": sandboxes,
        "in_flight": ctx.in_flight,
        "cap": cap,
        "remaining": max(cap - ctx.in_flight, 0),
    }


async def do_snapshot(ctx: HandlerCtx, payload: dict[str, Any]) -> dict[str, Any]:
    sandbox_id = _require_str(payload, "sandbox_id")
    snapshot_id = await ctx.client.snapshot(sandbox_id)
    return {"snapshot_id": snapshot_id}


async def do_expose_port(ctx: HandlerCtx, payload: dict[str, Any]) -> dict[str, Any]:
    sandbox_id = _require_str(payload, "sandbox_id")
    port = payload.get("port")
    if not isinstance(port, int) or port < 1 or port > 65535:
        raise SandboxError(S_PROVIDER_UNAVAILABLE, "invalid port")
    url = await ctx.client.expose_port(sandbox_id, port)
    return {"url": url}
